crypto/s2n_pkey.c

```diff
@@ -0,0 +1,187 @@
+{
+    notnull_check(pkey->decrypt);
+
+    return pkey->decrypt(pkey, in, out);
+}
+
+int s2n_pkey_match(const struct s2n_pkey *pub_key, const struct s2n_pkey *priv_key)
+{
+    notnull_check(pub_key->match);
+
+    return pub_key->match(pub_key, priv_key);
+}
+
+int s2n_pkey_free(struct s2n_pkey *pkey)
+{
+    if (pkey == NULL || pkey->free == NULL) {
+        return 0;
+    }
+    return pkey->free(pkey);
+}
+
+int s2n_asn1der_to_private_key(struct s2n_pkey *priv_key, struct s2n_blob *asn1der)
+{
+    uint8_t *key_to_parse = asn1der->data;
+
+    /* Detect key type */
+    EVP_PKEY *evp_private_key = d2i_AutoPrivateKey(NULL, (const unsigned char **)(void *)&key_to_parse, asn1der->size);
+    if (evp_private_key == NULL) {
+        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
+    }
+    
+    /* If key parsing is successful, d2i_AutoPrivateKey increments *key_to_parse to the byte following the parsed data */
+    uint32_t parsed_len = key_to_parse - asn1der->data;
+    if (parsed_len != asn1der->size) {
+        EVP_PKEY_free(evp_private_key);
+        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
+    }
+
+    /* Initialize s2n_pkey according to key type */
+    int type = EVP_PKEY_base_id(evp_private_key);
+    
+    int ret;
+    switch (type) {
+    case EVP_PKEY_RSA:
+        if ((ret = s2n_rsa_pkey_init(priv_key)) != 0) {
+            break;
+        }
+        ret = s2n_evp_pkey_to_rsa_private_key(&priv_key->key.rsa_key, evp_private_key);
+        break;
+    case EVP_PKEY_EC:
+        if ((ret = s2n_ecdsa_pkey_init(priv_key)) != 0) { 
+            break;
+        }
+        ret = s2n_evp_pkey_to_ecdsa_private_key(&priv_key->key.ecdsa_key, evp_private_key);
+        break;
+    default:
+        EVP_PKEY_free(evp_private_key);
+        S2N_ERROR(S2N_ERR_DECODE_PRIVATE_KEY);
+    }
+    
+    EVP_PKEY_free(evp_private_key);
+    
+    return ret;
+}
+
+int s2n_asn1der_to_public_key(struct s2n_pkey *pub_key, struct s2n_blob *asn1der)
+{
+    uint8_t *cert_to_parse = asn1der->data;
+    
+    X509 *cert = d2i_X509(NULL, (const unsigned char **)(void *)&cert_to_parse, asn1der->size);
+    if (cert == NULL) {
+        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
+    }
+    
+    /* If cert parsing is successful, d2i_X509 increments *cert_to_parse to the byte following the parsed data */
+    uint32_t parsed_len = cert_to_parse - asn1der->data;
+    if (parsed_len != asn1der->size) {
+        X509_free(cert);
+        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
+    }
+
+    EVP_PKEY *evp_public_key = X509_get_pubkey(cert);
+    X509_free(cert);
+
+    if (evp_public_key == NULL) {
+        S2N_ERROR(S2N_ERR_DECODE_CERTIFICATE);
+    }
+
+    /* Check for success in decoding certificate according to type */
+    int type = EVP_PKEY_base_id(evp_public_key);
+    
+    int ret;
+    switch (type) {
+    case EVP_PKEY_RSA:
+        if ((ret = s2n_rsa_pkey_init(pub_key)) != 0) {
```

Comment:
Let's not do assignment in the conditionals. In my opinion, adding an extra statement to set `ret` before the conditional will help readibility and reduce chance of someone introducing an error here later.